lib/p5.fab.js

```diff
@@ -258,376 +269,463 @@ class Fab {
-        // this if statement was added
-        if (e == null) {
-            this.e = this.makeE(x, y, z);
-        }
-        else if (multiplier) {
-            this.e = e * this.makeE(x, y, z);
+  render() {
+    if (this.coordinateSystem == "delta") {
+      this.drawDeltaPrinter();
+    } else {
+      this.drawCartesianPrinter();
+    }
+
+    if (!this.model) {
+      //tracks current toolpath position
+      var toolpathPos = new p5.Vector(0, 0, 0); // assume you're homed to start
+      beginShape(LINES);
+      for (let v in this.vertices) {
+        v = parseInt(v);
+        var vertexData = this.vertices[v];
+        if (vertexData.command == "G0") {
+          //move toolpath position
+          toolpathPos = toolpathPos.set([
+            vertexData.vertex.x,
+            vertexData.vertex.y,
+            vertexData.vertex.z,
+          ]);
+          continue; // no extrusions on G0
+        } else if (vertexData.command == "G1") {
+          // draw a line between current toolpath position and next toolpath position,
+          // sent toolpath position
+          vertex(toolpathPos.x, toolpathPos.y, toolpathPos.z);
+          vertex(vertexData.vertex.x, vertexData.vertex.y, vertexData.vertex.z);
+          toolpathPos = toolpathPos.set([
+            vertexData.vertex.x,
+            vertexData.vertex.y,
+            vertexData.vertex.z,
+          ]);
         }
-        else {
-            this.e = e;
+      }
+      endShape();
+      this.model = saveShape();
+    } else {
+      model(this.model);
+    }
+    pop();
+
+    // update camera position & orientation
+    if (_recoverCameraPosition) {
+      this.camera.setPosition(
+        this.cameraPosition.x,
+        this.cameraPosition.y,
+        this.cameraPosition.z
+      );
+      _recoverCameraPosition = false;
+      this.camera.lookAt(
+        this.cameraOrientation.x,
+        this.cameraOrientation.y,
+        this.cameraOrientation.z
+      );
+    }
+
+    this.cameraPosition.x = this.camera.eyeX;
+    this.cameraPosition.y = this.camera.eyeY;
+    this.cameraPosition.z = this.camera.eyeZ;
+    this.cameraOrientation.x = this.camera.centerX;
+    this.cameraOrientation.y = this.camera.centerY;
+    this.cameraOrientation.z = this.camera.centerZ;
+  }
+
+  drawCartesianPrinter() {
+    orbitControl(2, 2, 0.1);
+
+    // draw print bed
+    translate(-this.maxX / 2, 0, -this.maxY / 2);
+    rotateY(PI);
+    scale(-1, 1);
+    push();
+    translate(this.maxX / 2, 0, this.maxY / 2);
+    rotateY(PI / 12);
+    rotateX(PI / 12);
+    fill(254, 249, 152);
+    push();
+    translate(0, 2.5, 0);
+    box(this.maxX + 1, 5, this.maxY + 1); // build plate
+    pop();
+
+    push();
+    noFill();
+    translate(0, -this.maxZ / 2 + 1, 0);
+    stroke(220, 50, 32);
+    box(this.maxX, this.maxZ, this.maxY); // work envolope
+    pop();
+
+    noFill();
+    stroke(0);
+    translate(-this.maxX / 2, 0, -this.maxY / 2);
+  }
+
+  drawDeltaPrinter() {
```

Comment:
the differences are:
* using "radius" instead of "maxX/Y"
* using "cylinder" instead of "box" for bed
* putting the model in the middle (0,0)